taichi/rhi/CMakeLists.txt

```diff
@@ -0,0 +1,113 @@
+# ./taichi/rhi/CMakeLists.txt
+
+set(TAICHI_DEVICE_API ti_device_api)
+add_library(${TAICHI_DEVICE_API} STATIC)
+target_sources(${TAICHI_DEVICE_API}
+  PRIVATE
+    arch.cpp
+    device.cpp
+    device_capability.cpp
+    window_system.cpp
+  )
+
+target_include_directories(${TAICHI_DEVICE_API}
+  PRIVATE
+    ${PROJECT_SOURCE_DIR}
+  )
+
+if (TI_WITH_OPENGL OR TI_WITH_VULKAN AND NOT ANDROID)
+  set(GLFW_BUILD_DOCS OFF CACHE BOOL "" FORCE)
+  set(GLFW_BUILD_TESTS OFF CACHE BOOL "" FORCE)
+  set(GLFW_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+
+  if (APPLE)
+    set(GLFW_VULKAN_STATIC ON CACHE BOOL "" FORCE)
+  endif()
+
+  message("Building with GLFW")
+  # set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_GLFW")
+  target_compile_definitions(${TAICHI_DEVICE_API} PUBLIC TI_WITH_GLFW)
+  add_subdirectory(${PROJECT_SOURCE_DIR}/external/glfw ${PROJECT_BINARY_DIR}/external/glfw)
+  target_link_libraries(${TAICHI_DEVICE_API} PRIVATE glfw)
+  target_include_directories(${TAICHI_DEVICE_API} SYSTEM
+    PUBLIC
+      ${PROJECT_SOURCE_DIR}/external/glfw/include
+    )
+endif()
+
+target_include_directories(${TAICHI_DEVICE_API} SYSTEM
+    PRIVATE
+      ${PROJECT_SOURCE_DIR}/external/spdlog/include
+    )
+
+if (TI_WITH_METAL)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_METAL")
+    add_subdirectory(metal)
+    target_link_libraries(${TAICHI_DEVICE_API} PRIVATE metal_rhi)
+endif()
+
+if (TI_WITH_OPENGL)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_OPENGL")
+    add_subdirectory(opengl)
+    target_link_libraries(${TAICHI_DEVICE_API} PRIVATE opengl_rhi)
+endif()
+
+if (TI_WITH_DX11)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_DX11")
+    add_subdirectory(dx)
+    target_link_libraries(${TAICHI_DEVICE_API} PRIVATE dx_rhi)
+endif()
+
+if (TI_WITH_VULKAN)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_VULKAN")
+    if (APPLE)
+        # The latest Molten-vk v1.2.0 and v1.1.11 breaks GGUI: mpm3d_ggui.py
+        # So we have to manually download and install Molten-vk v1.10.0
+        #
+        # Uncomment the following lines if the mpm3d_ggui.py runs well with the latest Molten-vk
+        #find_library(MOLTEN_VK libMoltenVK.dylib PATHS $HOMEBREW_CELLAR/molten-vk $VULKAN_SDK REQUIRED)
+        #configure_file(${MOLTEN_VK} ${CMAKE_BINARY_DIR}/libMoltenVK.dylib COPYONLY)
+        #message(STATUS "MoltenVK library ${MOLTEN_VK}")
+
+        if(NOT EXISTS ${CMAKE_BINARY_DIR}/libMoltenVK.dylib)
```

Comment:
OOC: is GGUI the only reason that we need molten-vk on Mac?

Reply:
We still have vulkan compute backend

Reply:
aha right, anything blocking us from removing vulkan backend on mac in favor of metal btw?